tfe/resource_tfe_team_project_access.go

```diff
@@ -166,3 +448,23 @@ func resourceTFETeamProjectAccessDelete(ctx context.Context, d *schema.ResourceD
 
 	return nil
 }
+
+// You cannot set custom permissions when access level is not "custom"
+func checkForCustomPermissions(_ context.Context, d *schema.ResourceDiff, meta interface{}) error {
+
+	if access, ok := d.GetOk("access"); ok && access != "custom" {
+		// is an empty [] if project_access is not in the config
+		project_access := d.GetRawConfig().GetAttr("project_access").AsValueSet().Values()
+		if len(project_access) != 0 {
+			return fmt.Errorf("you can only set project_access permissions with access level custom")
+		}
+
+		// is an empty [] if project_access is not in the config
+		workspace_access := d.GetRawConfig().GetAttr("workspace_access").AsValueSet().Values()
+		if len(workspace_access) != 0 {
+			return fmt.Errorf("you can only set workspace_access permissions with access level custom")
+		}
+	}
+
+	return nil
+}
```

Comment:
Ultimately I'm calling this out because I _think_ there could be an edge case here. The scenario I'm thinking:

1. The resource is creating using the provider using `custom` permissions. 
2. Someone in the UI modifies the permissions from `custom` -> `read`
3. A second apply is attempted with no config changes.

Reply:
I just tested this out by:
* apply a custom config
* changing the access level to read in the ui
* applying the same config without making any changes to it -- so it's still custom and has all the stuff

The apply was successful:
```
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # tfe_team_project_access.custom_foobar will be updated in-place
  ~ resource "tfe_team_project_access" "custom_foobar" {
      ~ access     = "read" -> "custom"
        id         = "tprj-R7FXUdKiDKUp9XuR"
        # (2 unchanged attributes hidden)

      ~ project_access {
          ~ settings = "read" -> "delete"
          ~ teams    = "none" -> "manage"
        }

      ~ workspace_access {
          ~ create         = false -> true
          ~ delete         = false -> true
          ~ locking        = false -> true
          ~ move           = false -> true
          ~ run_tasks      = false -> true
          ~ runs           = "read" -> "apply"
          ~ variables      = "read" -> "write"
            # (2 unchanged attributes hidden)
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

tfe_team_project_access.custom_foobar: Modifying... [id=tprj-R7FXUdKiDKUp9XuR]
tfe_team_project_access.custom_foobar: Modifications complete after 0s [id=tprj-R7FXUdKiDKUp9XuR]
```

Is that expected or is it supposed to fail??

Reply:
I guess this makes sense. Terraform will detect drift, refresh state and the plan output will reflect what your config's intent is.